google-cloud-firestore/src/main/java/com/google/cloud/firestore/collection/ArraySortedMap.java

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore.collection;
+
+import com.google.api.core.InternalApi;
+import java.util.*;
+
+/**
+ * This is an array backed implementation of ImmutableSortedMap. It uses arrays and linear lookups
+ * to achieve good memory efficiency while maintaining good performance for small collections. To
+ * avoid degrading performance with increasing collection size it will automatically convert to a
+ * RBTreeSortedMap after an insert call above a certain threshold.
+ */
```

Comment:
If these were copied from somewhere else, could we include a link to the original implementation?

Reply:
They are copied out of Google3, but are available elsewhere. I added a note that points to the location in the Firebase Admin SDK.